inst/include/dust/dust.hpp

```diff
@@ -45,8 +45,93 @@ struct pars_t {
     pars_t(dust::nothing(), internal_) {
   }
 };
-}
 
+// If we store history, we need to save:
+//
+// history order (n_particles * n_data)
+// history value (n_history_state * n_particles * n_data)
+//
+// restart (not yet thinking about this)
+//
+// implies that we should save an index
+
+template <typename real_t>
+class filter_state {
+public:
+  filter_state(size_t n_state, size_t n_particles, size_t n_data) :
+    n_state_(n_state), n_particles_(n_particles), n_data_(n_data), offset_(0) {
+    resize(n_state, n_particles, n_data);
+  }
+
+  // default constructable
+  filter_state() : filter_state(0, 0, 0) {
+  }
+
+  void resize(size_t n_state, size_t n_particles, size_t n_data) {
+    n_state_ = n_state;
+    n_particles_ = n_particles;
+    n_data_ = n_data;
+    offset_ = 0;
+    history_value.resize(n_state_ * n_particles_ * (n_data_ + 1));
+    history_order.resize(n_particles_ * (n_data_ + 1));
+    for (size_t i = 0; i < n_particles_; ++i) {
+      history_order[i] = i;
+    }
+  }
+
+  typename std::vector<real_t>::iterator history_value_iterator() {
+    return history_value.begin() + offset_ * n_state_ * n_particles_;
+  }
+
+  typename std::vector<size_t>::iterator history_order_iterator() {
+    return history_order.begin() + offset_ * n_particles_;
+  }
+
+  std::vector<real_t> history() const {
+    std::vector<real_t> ret(size());
+    history(ret.begin());
+    return ret;
+  }
+
+  template <typename Iterator>
+  void history(Iterator ret) const {
+    std::vector<size_t> index_particle(n_particles_);
+    for (size_t i = 0; i < n_particles_; ++i) {
+      index_particle[i] = i;
+    }
+    for (size_t k = 0; k < n_data_ + 1; ++k) {
+      size_t i = n_data_ - k;
+      auto it_order = history_order.begin() + i * n_particles_;
+      auto it_value = history_value.begin() + i * n_state_ * n_particles_;
+      auto it_ret = ret + i * n_state_ * n_particles_;
+      for (size_t j = 0; j < n_particles_; ++j) {
+        const size_t idx = *(it_order + index_particle[j]);
+        index_particle[j] = idx;
+        std::copy_n(it_value + idx * n_state_, n_state_,
+                    it_ret + j * n_state_);
+      }
+    }
+  }
+
+  size_t size() const {
+    return history_value.size();
+  }
+
+  void advance() {
+    offset_++;
+  }
+
+  std::vector<real_t> history_value;
+  std::vector<size_t> history_order;
```

Comment:
Should these be private?

Reply:
yes, them being public was a left-over from a tedious bit of bookkeeping checking